

同学们好，本次作业的主要内容为使用pandas进行数据预处理。希望这两天你们已经从Python的课程内容中回过神来了。
没有数据的分析是无源之水，能够熟练处理数据是成为合格的数据分析师的基本要求，希望大家在今后多多实战，成为数据大师。

本次作业将使用公开标普500的股票数据。
https://www.kaggle.com/dgawlik/nyse#prices-split-adjusted.csv

作业的形式以问答为主，因为不涉及过长的代码，核心是数据的操作，所以这里没有太多的代码模板可供大家参考。
希望大家能从搜索引擎（google/bing）问答社区（stackoverflow）或者[官方文档](https://pandas.pydata.org/pandas-docs/stable/user_guide/index.html)或者我提供的jupyter notebooks 中找到回答问题需要的方法。
请时刻牢记数据的增删改查四种原子操作，思考我们的问题可以被如何分解。

那么首先，我们要import pandas

In [1]:
import pandas as pd
import os
path1 = r'./fundamentals.csv'
print(os.path.abspath(path1))
df = pd.read_csv(path1,index_col='Period Ending')
#print( [column for column in df] )
#print(df[7:22]['Net Income'])
print(df.loc[['2015-12-31'],['Net Income']].mean())
print(df.loc[['2015-12-31'],['Net Income']].max()-df.loc[['2015-12-31'],['Net Income']].min())
# 反馈所在列 df[df['Net Income'].isin([5.963000e+09])]



C:\Users\aw101\fundamentals.csv
Net Income    1.447248e+09
dtype: float64
Net Income    4.797000e+10
dtype: float64


In [89]:
import datetime as dt
#parser = lambda date: pd.datetime.strptime(date,'%d%b%Y') #字符串转换为时间变量
df2 =  pd.read_csv(path1,parse_dates=[2])
df2 = pd.DataFrame(df2,columns = ['Period Ending','Fixed Assets','Total Assets','Ticker Symbol'])
df2 = df2.set_index('Period Ending')
df2 = df2['2016']
df2 = df2[df2['Fixed Assets']>0]   #去除资产为0的数字

df2['assetratio'] = (df2['Fixed Assets']) / (df2['Total Assets']) 
print(df2['assetratio'].mean())
_min = df2['assetratio'].min()
print(_min)
df2[df2['assetratio'].isin([_min])]['Ticker Symbol']
#print(df2.head(9))


0.24652320884591447
0.0030655972015790236


Period Ending
2016-12-31    PFG
Name: Ticker Symbol, dtype: object

# 1. 从fundemantals.csv开始！

fundemantals.csv 是这些股票的年报数据

请用数据回答以下问题：

1. S&P500股票在2015年`net income`的均值是多少？最大值比最小值多多少？（每问10分，共计20分）

 2015:
  Mean:1.447248e+09
  Diff:4.797000e+10
 

2. S&P500股票在2016年的固定资产（fixed assets）占总资产(total assets)比例的均值是多少？固定资产占总资产比例最小的股票是的代码（ticker symbol）是什么？（每问10分，共计20分）

0.24652320884591447
0.0030655972015790236
Period Ending
2016-12-31    PFG


In [51]:
import pandas as pd
path2 = './securities.csv'
dfs = pd.read_csv(path2,parse_dates=[6])
dfs = pd.DataFrame(dfs,columns = ['Ticker symbol','GICS Sector','Address of Headquarters','Date first added'])
dfs = dfs.dropna(subset=['GICS Sector','Date first added'])
dfs = dfs.sort_values(by = ['Date first added'])
_group = dfs.loc[:,['Ticker symbol','Date first added']].groupby(dfs['GICS Sector']) 

print(u'最早')
print('*'*18)
for name,group in _group:
    print(name)    
    print(group.iloc[0])
    
print('/n'*3)
print(u'最晚') 
print('*'*18)
dfs = dfs.sort_values(by=['Date first added'],ascending = True )
_group2 = dfs.loc[:,['Ticker symbol','Date first added']].groupby(dfs['Address of Headquarters'])
for name2,group2 in _group2:
    print(name2)
    print(group2.iloc[0])
print(group2.shape)

#print(dfs.head(20))
#print([column for column in dfs])
#print(dfs.head())
#print(dfs.shape)

最早
******************
Consumer Discretionary
Ticker symbol                       MCD
Date first added    1970-06-30 00:00:00
Name: 296, dtype: object
Consumer Staples
Ticker symbol                       GIS
Date first added    1969-03-31 00:00:00
Name: 207, dtype: object
Energy
Ticker symbol                       SLB
Date first added    1965-03-31 00:00:00
Name: 399, dtype: object
Financials
Ticker symbol                       JPM
Date first added    1975-06-30 00:00:00
Name: 255, dtype: object
Health Care
Ticker symbol                       ABT
Date first added    1964-03-31 00:00:00
Name: 1, dtype: object
Industrials
Ticker symbol                       HON
Date first added    1964-03-31 00:00:00
Name: 230, dtype: object
Information Technology
Ticker symbol                       HPQ
Date first added    1974-12-31 00:00:00
Name: 233, dtype: object
Materials
Ticker symbol                       SHW
Date first added    1964-06-30 00:00:00
Name: 404, dtype: object
Real Estate
Ticker symbol

# 2. 加入securities.csv~

securities.csv包含了这些股票的基本信息

1. 请列举出各个sector中的加入时间最早的股票名称（10分）
2. 请列举出每一个州中加入时间最晚的股票名称（10分）


In [51]:
import pandas as pd

def top(df, n=3, column='Research and Development'):
    return df.sort_values(by=column)[-n:]

path1 = r'./fundamentals.csv'
path2 = './securities.csv'
data1 = pd.read_csv(path1,parse_dates=[2])
data2 = pd.read_csv(path2,parse_dates=[6])
#按代码作为key进行合并
data3 = pd.merge(data1,data2,left_on = 'Ticker Symbol',right_on = 'Ticker symbol',how = 'outer')
#_group3 = data3.groupby([data3['GICS Sector'],data3['Period Ending']])

data3 = data3.set_index('Period Ending')
data3 = data3['2013':'2016']
#print(data3.columns)
_group3 = data3['Research and Development'].groupby(data3['GICS Sector']).sum()

print(_group3)

_group4 = data3['Research and Development'].groupby([data3['GICS Sector'],data3['Ticker Symbol']]).sum().reset_index()#返回为dataframe

_group4 = _group4[_group4['Research and Development']!=0] # 去除 0 数据行

#_group4 = _group4.sort_values('Research and Development',ascending = False)  不需要排序

_group4 = _group4.groupby(_group4['GICS Sector'])

_group4.apply(top)

#for name,group in _group4:
#    print(name)
#    print(group)


#print(data3.groupby([data3['GICS Sector'],data3['Period Ending']]).sum())





GICS Sector
Consumer Discretionary         1.860178e+09
Consumer Staples               7.150872e+09
Energy                         2.076513e+09
Financials                     2.401300e+08
Health Care                    1.688513e+11
Industrials                    5.533400e+10
Information Technology         3.158830e+11
Materials                      2.061621e+10
Real Estate                    1.162100e+08
Telecommunications Services    0.000000e+00
Utilities                      0.000000e+00
Name: Research and Development, dtype: float64


GICS Sector Ticker Symbol  \
GICS Sector                                                        
Consumer Discretionary 27   Consumer Discretionary           HAS   
                       24   Consumer Discretionary          GRMN   
Consumer Staples       85         Consumer Staples           CPB   
                       83         Consumer Staples           CLX   
                       102        Consumer Staples            PG   
Energy                 123                  Energy            HP   
                       113                  Energy           BHI   
Financials             190              Financials           USB   
Health Care            199             Health Care          AMGN   
                       227             Health Care           MRK   
                       231             Health Care           PFE   
Industrials            251             Industrials           CAT   
                       302             Industrials           UTX   
                       250             Industrials            BA   
Information Technology 306  Information Technology          AAPL   
                       343  Information Technology          MSFT   
                       335  Information Technology          INTC   
Materials              385               Materials           PPG   
                       381               Materials           MON   
                       372               Materials            DD   
Real Estate            415             Real Estate           VNO   
                       417             Real Estate            WY   

                            Research and Development  
GICS Sector                                           
Consumer Discretionary 27               6.730910e+08  
                       24               1.187087e+09  
Consumer Staples       85               4.910000e+08  
                       83               5.320000e+08  
                       102              5.804172e+09  
Energy                 123              5.751300e+07  
                       113              2.019000e+09  
Financials             190              2.401300e+08  
Health Care            199              1.629000e+10  
                       227              2.138700e+10  
                       231              2.276100e+10  
Industrials            251              8.496000e+09  
                       302              9.433000e+09  
                       250              1.407600e+10  
Information Technology 306              2.862800e+10  
                       343              4.582600e+10  
                       335              4.701600e+10  
Materials              385              1.888000e+09  
                       381              6.350000e+09  
                       372              7.534000e+09  
Real Estate            415              3.221000e+07  
                       417              8.400000e+07

# 3. merge!

现在你需要同时处理来自两个表中的信息了

1. 请思考，合并两个表的信息的时候，我们应该用什么样的准则对其它们（10分）
2. 请列举每个sector在2013-2016年累计Research&Development的总投入（10分）
3. 请列举出每个sector中，在2013-2016年累计Research&development投入最大的3家公司的名称以及投入的数值（20分）

In [ ]:
来不及 做了 4

# 4. 现在让我们来看看更加复杂的数据

请导入price.csv，然后结合你的聪明才智回答以下问题（附加题，40分）

假设你是某基金公司的老板，现在对于每只股票，你都专门安排了一位负责它的交易员。公司规定每一位交易员手中的资金要么全部买入要么全部卖出（空仓，转化为现金）。假设2016年每一位交易员手中都有10000美元，假设他们都能够看到2016年全年的数据，假设他们都能抓住每一次机会，那么请问2016年底时，赚钱最多的股票是哪一只，赚了多少钱？